### PR TITLE
Andi/farmprofiledata

### PR DESCRIPTION
**Pull request by andi-halim**
[//]: # "These comments meant for your reference, they are invisible and don't need to be deleted"

## What's new in this PR?

[//]: # '####### YOUR ANSWER BELOW ###########'
Implemented the TopItemsGraph and RecentHarvestsGraph.
https://www.notion.so/calblueprint/Data-into-Top-5-Items-Recent-Harvests-Graphs-87da0fabb1324f0f933e14bedc5f7125

They are functional as shown in http://localhost:3000/farm/recYZO7NOT3NSGKCC (Nick's farm) !!
- TopItems groups all harvest logs together by item, sorts them all and finds the top 5. Displays them in descending order as well.
- RecentHarvests takes the last 9 months and displays their total pounds per month in a graph in order from 9 months ago to the current month. 

I tested some more on another farm and found issues with this farm. http://localhost:3000/farm/ecddezeBohwFjyuK
   - it has something to do with not deleting an index of the list if it is a duplicate? though it doesn't have problems in Nick's farm.
   - additionally if I console.log, the same sized "dictionary" always outputs through each iteration which is confusing. 

See https://airtable.com/tblvHdssj1Nh7iMKe/viwJDFBMSq10iR4ad?blocks=hide for the airtable. 

### How to Review

[//]: # 'The order in which to review files and 
what to expect when testing locally'
[//]: # '####### YOUR ANSWER BELOW ###########'
A farm prop is passed down from FarmProfile -> FarmGraphsTable -> Graph -> RecentHarvestsGraph/TopItemsGraph files. 
All code for the graphs are in RecentHarvestsGraph.js and TopItemsGraph.js

### Related PRs

[//]: # "Optional - related PRs you're waiting on
/ PRs that will conflict, etc"
[//]: # '####### YOUR ANSWER BELOW ###########'
[//]: # '############## END ##################'

### Next steps

[//]: # "What doesn't work yet, what's NOT in this 
PR/has to be done "
[//]: # '####### YOUR ANSWER BELOW ###########'
Debugging. 
Working on filtering with the graphs. Should be relatively easy for RecentHarvest and rather a bit difficult for Top Items.

### Screenshots

[//]: # '#### YOUR SCREENSHOTS BELOW ########'
image
image

### Design Status

[//]: # 'If this is a frontend PR, what is the expected 
status of the UI in this PR (lo, mid, high- fi?'
[//]: # '####### LOFI/MIDFI/HIFI ###########'
[//]: # '############## END ##################'

CC: @phoebeli23

### CODE REVIEW

**Comment by andi-halim** [comment]
After further testing, I think the bug may have something to do with when a new Airtable record introduces a new crop. 
- will change when I can this weekend

**Comment by phoebeli23** [comment]
Super thorough work Andi! Could you briefly document your code in RecentHarvestsGraph.js and TopItemsGraph.js? A high level overview of your algorithm would be very helpful for future reference. 

I am unable to see the farm you said was having issues. But here are the bugs that I found:

- Nick's Corn Farm: console error
Screen Shot 2021-05-05 at 5 32 59 PM
- Sandra's Farm: error upon page load
Screen Shot 2021-05-05 at 5 32 59 PM
- Sweet Tomatoes: error upon page load
Screen Shot 2021-05-05 at 5 48 17 PM

File: src/screens/shared/graphs/RecentHarvestsGraph.js [hunk 1]
```diff
@@ -1,19 +1,81 @@
...
+    const recentDates = getPrevMonths(9);
+    const recentValues = [0, 0, 0, 0, 0, 0, 0, 0, 0];
+    for (let i = 0; i < dict.length; i += 1) {
+      for (let j = 0; j < recentDates.length; j += 1) {
+        if (recentDates[j] === dict[i][0]) {
+          // eslint-disable-next-line prefer-destructuring
+          recentValues[j] = dict[i][1];
+        }
+      }
+    }
```
Comment (phoebeli23) [inline src/screens/shared/graphs/RecentHarvestsGraph.js]: Could you add some comments explaining each chunk of code here?

File: src/screens/shared/graphs/TopItemsGraph.js [hunk 2]
```diff
@@ -1,17 +1,91 @@
...
+      cropsStr += crops;
+      quantitiesStr += quantities;
+    }
+    console.log(cropsStr);
+    console.log(quantitiesStr);
+    const quantitiesFloats =
+      quantitiesStr === ''
+        ? []
+        : quantitiesStr.match(/\d+(?:\.\d+)?/g).map(Number);
+    this.setState({ cropsStr, quantitiesFloats });
```
Comment (phoebeli23) [inline src/screens/shared/graphs/TopItemsGraph.js]: here as well

File: src/screens/shared/graphs/TopItemsGraph.js [hunk 3]
```diff
@@ -1,17 +1,91 @@
...
+  }
+
+  formulateData = (cropsStr, quantitiesFloats) => {
```
Comment (phoebeli23) [inline src/screens/shared/graphs/TopItemsGraph.js:43]: can you add a description for this method

File: src/screens/shared/graphs/TopItemsGraph.js [hunk 4]
```diff
@@ -1,17 +1,91 @@
...
+        dict[i] = [cropsSplit[i], quantitiesFloats[i]];
+      }
+    }
+
+    return {
+      cropsToQuantity: dict
+    };
+  };
+
+  sortData = cropsToQuantity => {
```
Comment (phoebeli23) [inline src/screens/shared/graphs/TopItemsGraph.js:69]: 1-sent description

**Comment by andi-halim** [comment]
Issues that I see (not from code in this PR, but instead the VictoryGraph implementation):
- If there are no harvest logs at all, the empty TopItems graph is a bit buggy in that it shows text when it shouldn't. 
image

- When there are <5 unique crops, the graph is pretty poor in that it is not evenly spaced out and proportional. 
image

**Comment by andi-halim** [comment]
> Issues that I see (not from code in this PR, but instead the VictoryGraph implementation):
> 
> * If there are no harvest logs at all, the empty TopItems graph is a bit buggy in that it shows text when it shouldn't.
>   image
> * When there are <5 unique crops, the graph is pretty poor in that it is not evenly spaced out and proportional.
>   image

1. Render Empty Graph graphic by putting a boolean prop for IsEmpty in FarmProfileGraph component. 
2. Try to fill in the empty values of the top 5 List with empty strings and 0 to try to make the graph more proportional.

**Comment by andi-halim** [comment]
looks much better but victorygraphs still iffy on that proportionality. Just render N/A with a bit of a trick. 
- Since categories can't have the same name (or else their values are just summed and combined), I added an additional whitespace at the end of the "N/A" for every N/A value. 

image

**Comment by phoebeli23** [comment]
WOW you really outdid yourself here Andi :)

- The "N/A trick" is a really interesting solution for balancing the graph, would be worth documenting in the code why it's needed -- but the output is beautiful

- EmptyGraph appears when needed

Will give a more in-depth review tomorrow double-checking the numbers and code

**Pull request merged**